In [ ]:
import pandas as pd

In [ ]:
spotify_df = pd.read_csv('universal_top_spotify_songs.csv')
spotify_df

spotify_id  \
0      0AYt6NMyyLd0rLuvr0UkMH   
1      2V2K1hzCtgj9xAnga9WUTy   
2      1CrwpdodnkXcjPSrg8kw9g   
3      6HmiaxYKrzq3sU2M0eFDg2   
4      4tsVMjM60RNTe9EV5oQ4sQ   
...                       ...   
29108  7iQXYTyuG13aoeHxGG28Nh   
29109  2HafqoJbgXdtjwCOvNEF14   
29110  56y1jOTK0XSvJzVv9vHQBK   
29111  7x9aauaA9cu6tyfpHnqDLo   
29112  4MjDJD8cW7iVeWInc2Bdyj   

                                                    name  \
0                              Slime You Out (feat. SZA)   
1                       Horns In The Sun - Thakzin Remix   
2      Gangnam Style (feat. DJ Maphorisa & Kabza De S...   
3                                             BETTER NOW   
4                                            Masithokoze   
...                                                  ...   
29108                                        PERRO NEGRO   
29109                                        Si No Estás   
29110                                 Paint The Town Red   
29111                Seven (feat. Latto) (Explicit Ver.)   
29112                                             MONACO   

                                                 artists  daily_rank  \
0                                             Drake, SZA          50   
1          DJ Kent, Thakzin, Mo-T, MÖRDA, Brenden Praise          49   
2      Mas Musiq, Daliwonga, DJ Maphorisa, Kabza De S...          48   
3                                                A-Reece          47   
4                                       DJ Stokie, Eemoh          46   
...                                                  ...         ...   
29108                                    Bad Bunny, Feid           5   
29109                                     iñigo quintero           4   
29110                                           Doja Cat           3   
29111                                   Jung Kook, Latto           2   
29112                                          Bad Bunny           1   

       daily_movement  weekly_movement country snapshot_date  popularity  \
0                   0              -28      ZA    2023-10-25          84   
1                 -11                1      ZA    2023-10-25          55   
2                 -15              -13      ZA    2023-10-25          55   
3                   1                3      ZA    2023-10-25          51   
4                   4                4      ZA    2023-10-25          47   
...               ...              ...     ...           ...         ...   
29108              45                0     NaN    2023-10-18          89   
29109              46                0     NaN    2023-10-18          95   
29110              47                0     NaN    2023-10-18          87   
29111              48                0     NaN    2023-10-18          98   
29112              49                0     NaN    2023-10-18          91   

       is_explicit  ...  key loudness mode  speechiness  acousticness  \
0             True  ...    5   -9.243    0       0.0502        0.5080   
1            False  ...    5   -8.262    0       0.0670        0.0633   
2            False  ...    2  -12.642    0       0.0560        0.0632   
3             True  ...    1   -7.524    1       0.1660        0.1300   
4            False  ...    6  -14.253    0       0.0617        0.0224   
...            ...  ...  ...      ...  ...          ...           ...   
29108         True  ...    5   -2.248    1       0.2620        0.0887   
29109        False  ...    5   -8.720    1       0.0285        0.8270   
29110         True  ...    2   -7.683    0       0.1940        0.2550   
29111         True  ...   11   -4.107    1       0.0434        0.3110   
29112         True  ...    4   -5.009    0       0.0680        0.1500   

       instrumentalness  liveness  valence    tempo  time_signature  
0              0.000000    0.2590    0.105   88.880               3  
1              0.693000    0.3040    0.210  118.024               4  
2              0.000135    0.0400   

In [ ]:
spotify_df.isnull()

spotify_id   name  artists  daily_rank  daily_movement  \
0           False  False    False       False           False   
1           False  False    False       False           False   
2           False  False    False       False           False   
3           False  False    False       False           False   
4           False  False    False       False           False   
...           ...    ...      ...         ...             ...   
29108       False  False    False       False           False   
29109       False  False    False       False           False   
29110       False  False    False       False           False   
29111       False  False    False       False           False   
29112       False  False    False       False           False   

       weekly_movement  country  snapshot_date  popularity  is_explicit  ...  \
0                False    False          False       False        False  ...   
1                False    False          False       False        False  ...   
2                False    False          False       False        False  ...   
3                False    False          False       False        False  ...   
4                False    False          False       False        False  ...   
...                ...      ...            ...         ...          ...  ...   
29108            False     True          False       False        False  ...   
29109            False     True          False       False        False  ...   
29110            False     True          False       False        False  ...   
29111            False     True          False       False        False  ...   
29112            False     True          False       False        False  ...   

         key  loudness   mode  speechiness  acousticness  instrumentalness  \
0      False     False  False        False         False             False   
1      False     False  False        False         False             False   
2      False     False  False        False         False             False   
3      False     False  False        False         False             False   
4      False     False  False        False         False             False   
...      ...       ...    ...          ...           ...               ...   
29108  False     False  False        False         False             False   
29109  False     False  False        False         False             False   
29110  False     False  False        False         False             False   
29111  False     False  False        False         False             False   
29112  False     False  False        False         False             False   

       liveness  valence  tempo  time_signature  
0         False    False  False           False  
1         False    False  False           False  
2         False    False  False           False  
3         False    False  False           False  
4         False    False  False           False  
...         ...      ...    ...             ...  
29108     False    False  False           False  
29109     False    False  False           False  
29110     False    False  False           False  
29111     False    False  False           False  
29112     False    False  False           False  

[29113 rows x 25 columns]

In [ ]:
spotify_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29113 entries, 0 to 29112
Data columns (total 25 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   spotify_id          29113 non-null  object 
 1   name                29112 non-null  object 
 2   artists             29112 non-null  object 
 3   daily_rank          29113 non-null  int64  
 4   daily_movement      29113 non-null  int64  
 5   weekly_movement     29113 non-null  int64  
 6   country             28713 non-null  object 
 7   snapshot_date       29113 non-null  object 
 8   popularity          29113 non-null  int64  
 9   is_explicit         29113 non-null  bool   
 10  duration_ms         29113 non-null  int64  
 11  album_name          29112 non-null  object 
 12  album_release_date  29112 non-null  object 
 13  danceability        29113 non-null  float64
 14  energy              29113 non-null  float64
 15  key                 29113 non-null  int64  
 16  loud

In [ ]:
spotify_clean_df = spotify_df.copy()
spotify_clean_df.dropna(inplace=True)
spotify_clean_df.reset_index(inplace=True, drop=True)

spotify_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28712 entries, 0 to 28711
Data columns (total 25 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   spotify_id          28712 non-null  object 
 1   name                28712 non-null  object 
 2   artists             28712 non-null  object 
 3   daily_rank          28712 non-null  int64  
 4   daily_movement      28712 non-null  int64  
 5   weekly_movement     28712 non-null  int64  
 6   country             28712 non-null  object 
 7   snapshot_date       28712 non-null  object 
 8   popularity          28712 non-null  int64  
 9   is_explicit         28712 non-null  bool   
 10  duration_ms         28712 non-null  int64  
 11  album_name          28712 non-null  object 
 12  album_release_date  28712 non-null  object 
 13  danceability        28712 non-null  float64
 14  energy              28712 non-null  float64
 15  key                 28712 non-null  int64  
 16  loud

dapat dilihat setelah melakukan dropna, data yang tersisa sekarang 28712, dimana jumlah null data yang dihapus sebanyak 401 baris data, dan sekarang data tidak mengandung null data

In [ ]:
spotify_clean_df.head()

spotify_id                                               name  \
0  0AYt6NMyyLd0rLuvr0UkMH                          Slime You Out (feat. SZA)   
1  2V2K1hzCtgj9xAnga9WUTy                   Horns In The Sun - Thakzin Remix   
2  1CrwpdodnkXcjPSrg8kw9g  Gangnam Style (feat. DJ Maphorisa & Kabza De S...   
3  6HmiaxYKrzq3sU2M0eFDg2                                         BETTER NOW   
4  4tsVMjM60RNTe9EV5oQ4sQ                                        Masithokoze   

                                             artists  daily_rank  \
0                                         Drake, SZA          50   
1      DJ Kent, Thakzin, Mo-T, MÖRDA, Brenden Praise          49   
2  Mas Musiq, Daliwonga, DJ Maphorisa, Kabza De S...          48   
3                                            A-Reece          47   
4                                   DJ Stokie, Eemoh          46   

   daily_movement  weekly_movement country snapshot_date  popularity  \
0               0              -28      ZA    2023-10-25          84   
1             -11                1      ZA    2023-10-25          55   
2             -15              -13      ZA    2023-10-25          55   
3               1                3      ZA    2023-10-25          51   
4               4                4      ZA    2023-10-25          47   

   is_explicit  ...  key loudness mode  speechiness  acousticness  \
0         True  ...    5   -9.243    0       0.0502        0.5080   
1        False  ...    5   -8.262    0       0.0670        0.0633   
2        False  ...    2  -12.642    0       0.0560        0.0632   
3         True  ...    1   -7.524    1       0.1660        0.1300   
4        False  ...    6  -14.253    0       0.0617        0.0224   

   instrumentalness  liveness  valence    tempo  time_signature  
0          0.000000    0.2590    0.105   88.880               3  
1          0.693000    0.3040    0.210  118.024               4  
2          0.000135    0.0400    0.905  111.010               4  
3          0.000031    0.1210    0.725   83.001               4  
4          0.155000    0.0154    0.355  113.006               4  

[5 rows x 25 columns]

In [ ]:
spotify_fillna_df = spotify_df.copy()
spotify_fillna_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29113 entries, 0 to 29112
Data columns (total 25 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   spotify_id          29113 non-null  object 
 1   name                29112 non-null  object 
 2   artists             29112 non-null  object 
 3   daily_rank          29113 non-null  int64  
 4   daily_movement      29113 non-null  int64  
 5   weekly_movement     29113 non-null  int64  
 6   country             28713 non-null  object 
 7   snapshot_date       29113 non-null  object 
 8   popularity          29113 non-null  int64  
 9   is_explicit         29113 non-null  bool   
 10  duration_ms         29113 non-null  int64  
 11  album_name          29112 non-null  object 
 12  album_release_date  29112 non-null  object 
 13  danceability        29113 non-null  float64
 14  energy              29113 non-null  float64
 15  key                 29113 non-null  int64  
 16  loud

In [ ]:
spotify_fillna_df.country.fillna('Tanpa Negara', inplace=True)
spotify_fillna_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29113 entries, 0 to 29112
Data columns (total 25 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   spotify_id          29113 non-null  object 
 1   name                29112 non-null  object 
 2   artists             29112 non-null  object 
 3   daily_rank          29113 non-null  int64  
 4   daily_movement      29113 non-null  int64  
 5   weekly_movement     29113 non-null  int64  
 6   country             29113 non-null  object 
 7   snapshot_date       29113 non-null  object 
 8   popularity          29113 non-null  int64  
 9   is_explicit         29113 non-null  bool   
 10  duration_ms         29113 non-null  int64  
 11  album_name          29112 non-null  object 
 12  album_release_date  29112 non-null  object 
 13  danceability        29113 non-null  float64
 14  energy              29113 non-null  float64
 15  key                 29113 non-null  int64  
 16  loud

In [ ]:
spotify_fillna_df.tail()

spotify_id                                 name  \
29108  7iQXYTyuG13aoeHxGG28Nh                          PERRO NEGRO   
29109  2HafqoJbgXdtjwCOvNEF14                          Si No Estás   
29110  56y1jOTK0XSvJzVv9vHQBK                   Paint The Town Red   
29111  7x9aauaA9cu6tyfpHnqDLo  Seven (feat. Latto) (Explicit Ver.)   
29112  4MjDJD8cW7iVeWInc2Bdyj                               MONACO   

                artists  daily_rank  daily_movement  weekly_movement  \
29108   Bad Bunny, Feid           5              45                0   
29109    iñigo quintero           4              46                0   
29110          Doja Cat           3              47                0   
29111  Jung Kook, Latto           2              48                0   
29112         Bad Bunny           1              49                0   

            country snapshot_date  popularity  is_explicit  ...  key loudness  \
29108  Tanpa Negara    2023-10-18          89         True  ...    5   -2.248   
29109  Tanpa Negara    2023-10-18          95        False  ...    5   -8.720   
29110  Tanpa Negara    2023-10-18          87         True  ...    2   -7.683   
29111  Tanpa Negara    2023-10-18          98         True  ...   11   -4.107   
29112  Tanpa Negara    2023-10-18          91         True  ...    4   -5.009   

      mode  speechiness  acousticness  instrumentalness  liveness  valence  \
29108    1       0.2620        0.0887          0.000022    0.1790    0.345   
29109    1       0.0285        0.8270          0.000000    0.1380    0.524   
29110    0       0.1940        0.2550          0.000004    0.1120    0.726   
29111    1       0.0434        0.3110          0.000000    0.0815    0.890   
29112    0       0.0680        0.1500          0.000402    0.5800    0.130   

         tempo  time_signature  
29108   96.057               4  
29109   98.224               4  
29110   99.974               4  
29111  124.997               4  
29112  139.056               4  

[5 rows x 25 columns]

In [ ]:
spotify_ffill_df = spotify_df.copy()

spotify_ffill_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29113 entries, 0 to 29112
Data columns (total 25 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   spotify_id          29113 non-null  object 
 1   name                29112 non-null  object 
 2   artists             29112 non-null  object 
 3   daily_rank          29113 non-null  int64  
 4   daily_movement      29113 non-null  int64  
 5   weekly_movement     29113 non-null  int64  
 6   country             28713 non-null  object 
 7   snapshot_date       29113 non-null  object 
 8   popularity          29113 non-null  int64  
 9   is_explicit         29113 non-null  bool   
 10  duration_ms         29113 non-null  int64  
 11  album_name          29112 non-null  object 
 12  album_release_date  29112 non-null  object 
 13  danceability        29113 non-null  float64
 14  energy              29113 non-null  float64
 15  key                 29113 non-null  int64  
 16  loud

In [ ]:
spotify_ffill_df[spotify_ffill_df.country.isnull()].head(30)

spotify_id  \
3590  4iZ4pt7kvcaH6Yo8UoZ4s2   
3591  7DSAEUvxU8FajXtRloy8M0   
3592  3hUxzQpSfdDqwM3ZTFQY0K   
3593  25OeKzcqakFPaJlXHPE5lm   
3594  1iLbUmb2b2btyAAcQ68a1p   
3595  1dGr1c8CrMLDpV6mPbImSI   
3596  4SW9gHnW8NfKOdqmh0ij45   
3597  2QjOHCTQ1Jl3zawyYOpxh6   
3598  0V3wPSX9ygBnCm8psDIegu   
3599  3JvKfv6T31zO0ini8iNItO   
3600  2LBqCSwhJGcFQeTHMVGwy3   
3601  6XbtvPmIpyCbjuT0e8cQtp   
3602  4KULAymBBJcPRpk1yO4dOG   
3603  5ixdmTpCmH5diBA1j154wN   
3604  4qSEvFGCpde73gqIuq3sho   
3605  1vYXt7VSjH9JIM5oRRo7vA   
3606  7aqfrAY2p9BUSiupwk3svU   
3607  1Qrg8KqiBpW07V7PNxwwwL   
3608  7MXVkk9YMctZqd1Srtv4MB   
3609  1p80LdxRV74UKvL8gnD7ky   
3610  4Dvkj6JhhA12EX05fT7y2e   
3611  4rXLjWdF2ZZpXCVTfWcshS   
3612  4lIxdJw6W3Fg4vUIYCB0S5   
3613  59NraMJsLaMCVtwXTSia8i   
3614  7mXuWTczZNxG5EDcjFEuJR   
3615  5Eh1nj7IjV9lwpcKAkidyY   
3616  6wf7Yu7cxBSPrRlWeSeK0Q   
3617  5PyDJG7SQRgWXefgexqIge   
3618  1odExI7RdWc4BT515LTAwj   
3619  1kuGVB7EU95pJObxwvfwKS   

                                                   name  \
3590                                             Snooze   
3591                                            Flowers   
3592                                             august   
3593                                       Primera Cita   
3594               Milo J: Bzrp Music Sessions, Vol. 57   
3595                                              Lover   
3596                                              Fruto   
3597                                    Sweater Weather   
3598                                          Anti-Hero   
3599                                       Another Love   
3600                                        Die For You   
3601                                           Columbia   
3602      I Remember Everything (feat. Kacey Musgraves)   
3603                                         UN PREVIEW   
3604                                             HIBIKI   
3605            Dance The Night - From Barbie The Album   
3606               First Person Shooter (feat. J. Cole)   
3607                                          Kill Bill   
3608                                            Starboy   
3609                                        Blank Space   
3610                                          As It Was   
3611                                           fukumean   
3612                                              Style   
3613                                              Prada   
3614                                          LADY GAGA   
3615                                  One Of Your Girls   
3616  What Was I Made For? [From The Motion Picture ...   
3617                                        Agora Hills   
3618                                           Daylight   
3619                                            vampire   

                                       artists  daily_rank  daily_movement  \
3590                                       SZA          50               0   
3591                               Miley Cyrus          49               1   
3592                              Taylor Swift          48               0   
3593                                Carin Leon          47               0   
3594                          Bizarrap, Milo j          46               0   
3595                              Taylor Swift          45              -3   
3596                          Bizarrap, Milo j          44              -3   
3597                         The Neighbourhood          43               2   
3598                              Taylor Swift          42              -4   
3599                                 Tom Odell          41               3   
3600                                The Weeknd          40               0   
3601                                   Quevedo          39              -3   
3602               Zach Bryan, Kacey Musgraves          38               5   
3603                                 Bad Bunny          37              -2   
3604                     

In [ ]:
spotify_ffill_df.iloc[3589:3630]

spotify_id  \
3589  3rUGC1vUpkDG9CZFHMur1t   
3590  4iZ4pt7kvcaH6Yo8UoZ4s2   
3591  7DSAEUvxU8FajXtRloy8M0   
3592  3hUxzQpSfdDqwM3ZTFQY0K   
3593  25OeKzcqakFPaJlXHPE5lm   
3594  1iLbUmb2b2btyAAcQ68a1p   
3595  1dGr1c8CrMLDpV6mPbImSI   
3596  4SW9gHnW8NfKOdqmh0ij45   
3597  2QjOHCTQ1Jl3zawyYOpxh6   
3598  0V3wPSX9ygBnCm8psDIegu   
3599  3JvKfv6T31zO0ini8iNItO   
3600  2LBqCSwhJGcFQeTHMVGwy3   
3601  6XbtvPmIpyCbjuT0e8cQtp   
3602  4KULAymBBJcPRpk1yO4dOG   
3603  5ixdmTpCmH5diBA1j154wN   
3604  4qSEvFGCpde73gqIuq3sho   
3605  1vYXt7VSjH9JIM5oRRo7vA   
3606  7aqfrAY2p9BUSiupwk3svU   
3607  1Qrg8KqiBpW07V7PNxwwwL   
3608  7MXVkk9YMctZqd1Srtv4MB   
3609  1p80LdxRV74UKvL8gnD7ky   
3610  4Dvkj6JhhA12EX05fT7y2e   
3611  4rXLjWdF2ZZpXCVTfWcshS   
3612  4lIxdJw6W3Fg4vUIYCB0S5   
3613  59NraMJsLaMCVtwXTSia8i   
3614  7mXuWTczZNxG5EDcjFEuJR   
3615  5Eh1nj7IjV9lwpcKAkidyY   
3616  6wf7Yu7cxBSPrRlWeSeK0Q   
3617  5PyDJG7SQRgWXefgexqIge   
3618  1odExI7RdWc4BT515LTAwj   
3619  1kuGVB7EU95pJObxwvfwKS   
3620  3nNmRE0DxHC6ZaKkrpUumS   
3621  3cDyocF0Ibdfs1SFQk5cLZ   
3622  5XeFesFbtLpXzIVDNQP22n   
3623  4bw8mcDUSRWfQo63ZTYRnU   
3624  6uIIdjYTxxpWOyWuVXrKQO   
3625  2YSzYUF3jWqb9YP9VXmpjE   
3626  7ABLbnD53cQK00mhcaOUVG   
3627  0rKWJnmo6Q0ovoPOLoLm0t   
3628  5RqSsdzTNPX1uzkmlHCFvK   
3629  5aIVCx5tnk0ntmdiinnYvw   

                                                   name  \
3589                                             greedy   
3590                                             Snooze   
3591                                            Flowers   
3592                                             august   
3593                                       Primera Cita   
3594               Milo J: Bzrp Music Sessions, Vol. 57   
3595                                              Lover   
3596                                              Fruto   
3597                                    Sweater Weather   
3598                                          Anti-Hero   
3599                                       Another Love   
3600                                        Die For You   
3601                                           Columbia   
3602      I Remember Everything (feat. Kacey Musgraves)   
3603                                         UN PREVIEW   
3604                                             HIBIKI   
3605            Dance The Night - From Barbie The Album   
3606               First Person Shooter (feat. J. Cole)   
3607                                          Kill Bill   
3608                                            Starboy   
3609                                        Blank Space   
3610                                          As It Was   
3611                                           fukumean   
3612                                              Style   
3613                                              Prada   
3614                                          LADY GAGA   
3615                                  One Of Your Girls   
3616  What Was I Made For? [From The Motion Picture ...   
3617                                        Agora Hills   
3618                                           Daylight   
3619                                            vampire   
3620                                               FINA   
3621                                           Lollipop   
3622                                   I Wanna Be Yours   
3623                                        Según Quién   
3624                                           Que Onda   
3625                                 IDGAF (feat. Yeat)   
3626                                               LALA   
3627                                           TOO MUCH   
3628                                              QLONA   
3629                                              Water   

                                       artists  daily_rank  daily_movement  \
3589                                Tate McRae           1               0   
3590                                  

In [ ]:
spotify_ffill_df.ffill(inplace=True)

In [ ]:
spotify_ffill_df.iloc[3589:3630]

spotify_id  \
3589  3rUGC1vUpkDG9CZFHMur1t   
3590  4iZ4pt7kvcaH6Yo8UoZ4s2   
3591  7DSAEUvxU8FajXtRloy8M0   
3592  3hUxzQpSfdDqwM3ZTFQY0K   
3593  25OeKzcqakFPaJlXHPE5lm   
3594  1iLbUmb2b2btyAAcQ68a1p   
3595  1dGr1c8CrMLDpV6mPbImSI   
3596  4SW9gHnW8NfKOdqmh0ij45   
3597  2QjOHCTQ1Jl3zawyYOpxh6   
3598  0V3wPSX9ygBnCm8psDIegu   
3599  3JvKfv6T31zO0ini8iNItO   
3600  2LBqCSwhJGcFQeTHMVGwy3   
3601  6XbtvPmIpyCbjuT0e8cQtp   
3602  4KULAymBBJcPRpk1yO4dOG   
3603  5ixdmTpCmH5diBA1j154wN   
3604  4qSEvFGCpde73gqIuq3sho   
3605  1vYXt7VSjH9JIM5oRRo7vA   
3606  7aqfrAY2p9BUSiupwk3svU   
3607  1Qrg8KqiBpW07V7PNxwwwL   
3608  7MXVkk9YMctZqd1Srtv4MB   
3609  1p80LdxRV74UKvL8gnD7ky   
3610  4Dvkj6JhhA12EX05fT7y2e   
3611  4rXLjWdF2ZZpXCVTfWcshS   
3612  4lIxdJw6W3Fg4vUIYCB0S5   
3613  59NraMJsLaMCVtwXTSia8i   
3614  7mXuWTczZNxG5EDcjFEuJR   
3615  5Eh1nj7IjV9lwpcKAkidyY   
3616  6wf7Yu7cxBSPrRlWeSeK0Q   
3617  5PyDJG7SQRgWXefgexqIge   
3618  1odExI7RdWc4BT515LTAwj   
3619  1kuGVB7EU95pJObxwvfwKS   
3620  3nNmRE0DxHC6ZaKkrpUumS   
3621  3cDyocF0Ibdfs1SFQk5cLZ   
3622  5XeFesFbtLpXzIVDNQP22n   
3623  4bw8mcDUSRWfQo63ZTYRnU   
3624  6uIIdjYTxxpWOyWuVXrKQO   
3625  2YSzYUF3jWqb9YP9VXmpjE   
3626  7ABLbnD53cQK00mhcaOUVG   
3627  0rKWJnmo6Q0ovoPOLoLm0t   
3628  5RqSsdzTNPX1uzkmlHCFvK   
3629  5aIVCx5tnk0ntmdiinnYvw   

                                                   name  \
3589                                             greedy   
3590                                             Snooze   
3591                                            Flowers   
3592                                             august   
3593                                       Primera Cita   
3594               Milo J: Bzrp Music Sessions, Vol. 57   
3595                                              Lover   
3596                                              Fruto   
3597                                    Sweater Weather   
3598                                          Anti-Hero   
3599                                       Another Love   
3600                                        Die For You   
3601                                           Columbia   
3602      I Remember Everything (feat. Kacey Musgraves)   
3603                                         UN PREVIEW   
3604                                             HIBIKI   
3605            Dance The Night - From Barbie The Album   
3606               First Person Shooter (feat. J. Cole)   
3607                                          Kill Bill   
3608                                            Starboy   
3609                                        Blank Space   
3610                                          As It Was   
3611                                           fukumean   
3612                                              Style   
3613                                              Prada   
3614                                          LADY GAGA   
3615                                  One Of Your Girls   
3616  What Was I Made For? [From The Motion Picture ...   
3617                                        Agora Hills   
3618                                           Daylight   
3619                                            vampire   
3620                                               FINA   
3621                                           Lollipop   
3622                                   I Wanna Be Yours   
3623                                        Según Quién   
3624                                           Que Onda   
3625                                 IDGAF (feat. Yeat)   
3626                                               LALA   
3627                                           TOO MUCH   
3628                                              QLONA   
3629                                              Water   

                                       artists  daily_rank  daily_movement  \
3589                                Tate McRae           1               0   
3590                                  

In [ ]:
spotify_interpolate_df = spotify_df.copy()

spotify_interpolate_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29113 entries, 0 to 29112
Data columns (total 25 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   spotify_id          29113 non-null  object 
 1   name                29112 non-null  object 
 2   artists             29112 non-null  object 
 3   daily_rank          29113 non-null  int64  
 4   daily_movement      29113 non-null  int64  
 5   weekly_movement     29113 non-null  int64  
 6   country             28713 non-null  object 
 7   snapshot_date       29113 non-null  object 
 8   popularity          29113 non-null  int64  
 9   is_explicit         29113 non-null  bool   
 10  duration_ms         29113 non-null  int64  
 11  album_name          29112 non-null  object 
 12  album_release_date  29112 non-null  object 
 13  danceability        29113 non-null  float64
 14  energy              29113 non-null  float64
 15  key                 29113 non-null  int64  
 16  loud

In [ ]:
spotify_interpolate_df.interpolate()

spotify_id  \
0      0AYt6NMyyLd0rLuvr0UkMH   
1      2V2K1hzCtgj9xAnga9WUTy   
2      1CrwpdodnkXcjPSrg8kw9g   
3      6HmiaxYKrzq3sU2M0eFDg2   
4      4tsVMjM60RNTe9EV5oQ4sQ   
...                       ...   
29108  7iQXYTyuG13aoeHxGG28Nh   
29109  2HafqoJbgXdtjwCOvNEF14   
29110  56y1jOTK0XSvJzVv9vHQBK   
29111  7x9aauaA9cu6tyfpHnqDLo   
29112  4MjDJD8cW7iVeWInc2Bdyj   

                                                    name  \
0                              Slime You Out (feat. SZA)   
1                       Horns In The Sun - Thakzin Remix   
2      Gangnam Style (feat. DJ Maphorisa & Kabza De S...   
3                                             BETTER NOW   
4                                            Masithokoze   
...                                                  ...   
29108                                        PERRO NEGRO   
29109                                        Si No Estás   
29110                                 Paint The Town Red   
29111                Seven (feat. Latto) (Explicit Ver.)   
29112                                             MONACO   

                                                 artists  daily_rank  \
0                                             Drake, SZA          50   
1          DJ Kent, Thakzin, Mo-T, MÖRDA, Brenden Praise          49   
2      Mas Musiq, Daliwonga, DJ Maphorisa, Kabza De S...          48   
3                                                A-Reece          47   
4                                       DJ Stokie, Eemoh          46   
...                                                  ...         ...   
29108                                    Bad Bunny, Feid           5   
29109                                     iñigo quintero           4   
29110                                           Doja Cat           3   
29111                                   Jung Kook, Latto           2   
29112                                          Bad Bunny           1   

       daily_movement  weekly_movement country snapshot_date  popularity  \
0                   0              -28      ZA    2023-10-25          84   
1                 -11                1      ZA    2023-10-25          55   
2                 -15              -13      ZA    2023-10-25          55   
3                   1                3      ZA    2023-10-25          51   
4                   4                4      ZA    2023-10-25          47   
...               ...              ...     ...           ...         ...   
29108              45                0     NaN    2023-10-18          89   
29109              46                0     NaN    2023-10-18          95   
29110              47                0     NaN    2023-10-18          87   
29111              48                0     NaN    2023-10-18          98   
29112              49                0     NaN    2023-10-18          91   

       is_explicit  ...  key loudness mode  speechiness  acousticness  \
0             True  ...    5   -9.243    0       0.0502        0.5080   
1            False  ...    5   -8.262    0       0.0670        0.0633   
2            False  ...    2  -12.642    0       0.0560        0.0632   
3             True  ...    1   -7.524    1       0.1660        0.1300   
4            False  ...    6  -14.253    0       0.0617        0.0224   
...            ...  ...  ...      ...  ...          ...           ...   
29108         True  ...    5   -2.248    1       0.2620        0.0887   
29109        False  ...    5   -8.720    1       0.0285        0.8270   
29110         True  ...    2   -7.683    0       0.1940        0.2550   
29111         True  ...   11   -4.107    1       0.0434        0.3110   
29112         True  ...    4   -5.009    0       0.0680        0.1500   

       instrumentalness  liveness  valence    tempo  time_signature  
0              0.000000    0.2590    0.105   88.880               3  
1              0.693000    0.3040    0.210  118.024               4  
2              0.000135    0.0400   

In [ ]:
spotify_clean_df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
28707    False
28708    False
28709    False
28710    False
28711    False
Length: 28712, dtype: bool

In [ ]:
spotify_clean_df.country.duplicated()

0        False
1         True
2         True
3         True
4         True
         ...  
28707     True
28708     True
28709     True
28710     True
28711     True
Name: country, Length: 28712, dtype: bool

In [ ]:
spotify_country = spotify_clean_df.country.drop_duplicates()
spotify_country.reset_index(drop= True, inplace= True)

spotify_country.to_csv('country.csv', index=False)

In [ ]:
spotify_clean_df.drop_duplicates(subset=['name', 'artists'])

spotify_id  \
0      0AYt6NMyyLd0rLuvr0UkMH   
1      2V2K1hzCtgj9xAnga9WUTy   
2      1CrwpdodnkXcjPSrg8kw9g   
3      6HmiaxYKrzq3sU2M0eFDg2   
4      4tsVMjM60RNTe9EV5oQ4sQ   
...                       ...   
24371  7E8LK4R9gAkSG4CD1kqvsX   
24671  04Rp6XjTowBcL7gIbJf7uc   
24721  5eID9KNVJV2fDSPQGzQHQg   
24725  0SWDToufOxRVIngk8c1oZi   
24973  5qEGAHerczFt3HmL9lhh4j   

                                                    name  \
0                              Slime You Out (feat. SZA)   
1                       Horns In The Sun - Thakzin Remix   
2      Gangnam Style (feat. DJ Maphorisa & Kabza De S...   
3                                             BETTER NOW   
4                                            Masithokoze   
...                                                  ...   
24371                                    Virtuální drogy   
24671                                             ВМЕСТЕ   
24721                              Oficializar - Ao Vivo   
24725                          AMBIENTE ERRADO - Ao Vivo   
24973                  Nachts wach (Lila Wolken Bootleg)   

                                                 artists  daily_rank  \
0                                             Drake, SZA          50   
1          DJ Kent, Thakzin, Mo-T, MÖRDA, Brenden Praise          49   
2      Mas Musiq, Daliwonga, DJ Maphorisa, Kabza De S...          48   
3                                                A-Reece          47   
4                                       DJ Stokie, Eemoh          46   
...                                                  ...         ...   
24371                                Viktor Sheen, Calin          50   
24671                                 Aarne, BUSHIDO ZHO          50   
24721                    Gusttavo Lima, Maiara & Maraisa          50   
24725                                       Luan Santana          46   
24973                             Miksu / Macloud, makko          48   

       daily_movement  weekly_movement country snapshot_date  popularity  \
0                   0              -28      ZA    2023-10-25          84   
1                 -11                1      ZA    2023-10-25          55   
2                 -15              -13      ZA    2023-10-25          55   
3                   1                3      ZA    2023-10-25          51   
4                   4                4      ZA    2023-10-25          47   
...               ...              ...     ...           ...         ...   
24371               0                0      CZ    2023-10-19          58   
24671               0                0      BY    2023-10-19          63   
24721               0                0      BR    2023-10-19          81   
24725               0                4      BR    2023-10-19          67   
24973               0                2      AT    2023-10-19          74   

       is_explicit  ...  key loudness mode  speechiness  acousticness  \
0             True  ...    5   -9.243    0       0.0502        0.5080   
1            False  ...    5   -8.262    0       0.0670        0.0633   
2            False  ...    2  -12.642    0       0.0560        0.0632   
3             True  ...    1   -7.524    1       0.1660        0.1300   
4            False  ...    6  -14.253    0       0.0617        0.0224   
...            ...  ...  ...      ...  ...          ...           ...   
24371         True  ...   10   -8.070    0       0.0722        0.2200   
24671         True  ...    0   -6.382    1       0.0543        0.3680   
24721        False  ...   11   -4.761    1       0.0478        0.3430   
24725        False  ...    2   -5.347    1       0.0378        0.0368   
24973         True  ...   10   -4.400    0       0.1240        0.1650   

       instrumentalness  liveness  valence    tempo  time_signature  
0              0.000000    0.2590    0.105   88.880               3  
1              0.693000    0.3040    0.210  118.024               4  
2              0.000135    0.0400   

In [ ]:
spotify_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28712 entries, 0 to 28711
Data columns (total 25 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   spotify_id          28712 non-null  object 
 1   name                28712 non-null  object 
 2   artists             28712 non-null  object 
 3   daily_rank          28712 non-null  int64  
 4   daily_movement      28712 non-null  int64  
 5   weekly_movement     28712 non-null  int64  
 6   country             28712 non-null  object 
 7   snapshot_date       28712 non-null  object 
 8   popularity          28712 non-null  int64  
 9   is_explicit         28712 non-null  bool   
 10  duration_ms         28712 non-null  int64  
 11  album_name          28712 non-null  object 
 12  album_release_date  28712 non-null  object 
 13  danceability        28712 non-null  float64
 14  energy              28712 non-null  float64
 15  key                 28712 non-null  int64  
 16  loud

In [ ]:
spotify_clean_df.duration_ms.astype('float64')

0        310490.0
1        336657.0
2        350309.0
3        163435.0
4        427217.0
           ...   
28707    200373.0
28708    230480.0
28709    172964.0
28710    184400.0
28711    131872.0
Name: duration_ms, Length: 28712, dtype: float64

In [ ]:
spotify_clean_df.duration_ms = spotify_clean_df.duration_ms.astype('float64')

In [ ]:
spotify_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28712 entries, 0 to 28711
Data columns (total 25 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   spotify_id          28712 non-null  object 
 1   name                28712 non-null  object 
 2   artists             28712 non-null  object 
 3   daily_rank          28712 non-null  int64  
 4   daily_movement      28712 non-null  int64  
 5   weekly_movement     28712 non-null  int64  
 6   country             28712 non-null  object 
 7   snapshot_date       28712 non-null  object 
 8   popularity          28712 non-null  int64  
 9   is_explicit         28712 non-null  bool   
 10  duration_ms         28712 non-null  float64
 11  album_name          28712 non-null  object 
 12  album_release_date  28712 non-null  object 
 13  danceability        28712 non-null  float64
 14  energy              28712 non-null  float64
 15  key                 28712 non-null  int64  
 16  loud

In [ ]:
spotify_clean_df.snapshot_date = spotify_clean_df.snapshot_date.astype('datetime64')

In [ ]:
spotify_clean_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28712 entries, 0 to 28711
Data columns (total 25 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   spotify_id          28712 non-null  object        
 1   name                28712 non-null  object        
 2   artists             28712 non-null  object        
 3   daily_rank          28712 non-null  int64         
 4   daily_movement      28712 non-null  int64         
 5   weekly_movement     28712 non-null  int64         
 6   country             28712 non-null  object        
 7   snapshot_date       28712 non-null  datetime64[ns]
 8   popularity          28712 non-null  int64         
 9   is_explicit         28712 non-null  bool          
 10  duration_ms         28712 non-null  float64       
 11  album_name          28712 non-null  object        
 12  album_release_date  28712 non-null  object        
 13  danceability        28712 non-null  float64   

In [ ]:
spotify_clean_df.head()

spotify_id                                               name  \
0  0AYt6NMyyLd0rLuvr0UkMH                          Slime You Out (feat. SZA)   
1  2V2K1hzCtgj9xAnga9WUTy                   Horns In The Sun - Thakzin Remix   
2  1CrwpdodnkXcjPSrg8kw9g  Gangnam Style (feat. DJ Maphorisa & Kabza De S...   
3  6HmiaxYKrzq3sU2M0eFDg2                                         BETTER NOW   
4  4tsVMjM60RNTe9EV5oQ4sQ                                        Masithokoze   

                                             artists  daily_rank  \
0                                         Drake, SZA          50   
1      DJ Kent, Thakzin, Mo-T, MÖRDA, Brenden Praise          49   
2  Mas Musiq, Daliwonga, DJ Maphorisa, Kabza De S...          48   
3                                            A-Reece          47   
4                                   DJ Stokie, Eemoh          46   

   daily_movement  weekly_movement country snapshot_date  popularity  \
0               0              -28      ZA    2023-10-25          84   
1             -11                1      ZA    2023-10-25          55   
2             -15              -13      ZA    2023-10-25          55   
3               1                3      ZA    2023-10-25          51   
4               4                4      ZA    2023-10-25          47   

   is_explicit  ...  key loudness mode  speechiness  acousticness  \
0         True  ...    5   -9.243    0       0.0502        0.5080   
1        False  ...    5   -8.262    0       0.0670        0.0633   
2        False  ...    2  -12.642    0       0.0560        0.0632   
3         True  ...    1   -7.524    1       0.1660        0.1300   
4        False  ...    6  -14.253    0       0.0617        0.0224   

   instrumentalness  liveness  valence    tempo  time_signature  
0          0.000000    0.2590    0.105   88.880               3  
1          0.693000    0.3040    0.210  118.024               4  
2          0.000135    0.0400    0.905  111.010               4  
3          0.000031    0.1210    0.725   83.001               4  
4          0.155000    0.0154    0.355  113.006               4  

[5 rows x 25 columns]

In [ ]:
spotify_clean_df.name = spotify_clean_df.name.str.strip()

spotify_clean_df

spotify_id  \
0      0AYt6NMyyLd0rLuvr0UkMH   
1      2V2K1hzCtgj9xAnga9WUTy   
2      1CrwpdodnkXcjPSrg8kw9g   
3      6HmiaxYKrzq3sU2M0eFDg2   
4      4tsVMjM60RNTe9EV5oQ4sQ   
...                       ...   
28707  3xMHXmedL5Rvfxmiar9Ryv   
28708  56y1jOTK0XSvJzVv9vHQBK   
28709  5mjYQaktjmjcMKcUIcqz4s   
28710  7x9aauaA9cu6tyfpHnqDLo   
28711  3rUGC1vUpkDG9CZFHMur1t   

                                                    name  \
0                              Slime You Out (feat. SZA)   
1                       Horns In The Sun - Thakzin Remix   
2      Gangnam Style (feat. DJ Maphorisa & Kabza De S...   
3                                             BETTER NOW   
4                                            Masithokoze   
...                                                  ...   
28707                                            Chaleya   
28708                                 Paint The Town Red   
28709                                          Strangers   
28710                Seven (feat. Latto) (Explicit Ver.)   
28711                                             greedy   

                                                 artists  daily_rank  \
0                                             Drake, SZA          50   
1          DJ Kent, Thakzin, Mo-T, MÖRDA, Brenden Praise          49   
2      Mas Musiq, Daliwonga, DJ Maphorisa, Kabza De S...          48   
3                                                A-Reece          47   
4                                       DJ Stokie, Eemoh          46   
...                                                  ...         ...   
28707  Anirudh Ravichander, Arijit Singh, Shilpa Rao,...           5   
28708                                           Doja Cat           4   
28709                                        Kenya Grace           3   
28710                                   Jung Kook, Latto           2   
28711                                         Tate McRae           1   

       daily_movement  weekly_movement country snapshot_date  popularity  \
0                   0              -28      ZA    2023-10-25          84   
1                 -11                1      ZA    2023-10-25          55   
2                 -15              -13      ZA    2023-10-25          55   
3                   1                3      ZA    2023-10-25          51   
4                   4                4      ZA    2023-10-25          47   
...               ...              ...     ...           ...         ...   
28707              45                0      AE    2023-10-18          84   
28708              46                0      AE    2023-10-18          87   
28709              47                0      AE    2023-10-18          97   
28710              48                0      AE    2023-10-18          98   
28711              49                0      AE    2023-10-18          98   

       is_explicit  ...  key loudness mode  speechiness  acousticness  \
0             True  ...    5   -9.243    0       0.0502        0.5080   
1            False  ...    5   -8.262    0       0.0670        0.0633   
2            False  ...    2  -12.642    0       0.0560        0.0632   
3             True  ...    1   -7.524    1       0.1660        0.1300   
4            False  ...    6  -14.253    0       0.0617        0.0224   
...            ...  ...  ...      ...  ...          ...           ...   
28707        False  ...    0   -5.280    1       0.0852        0.3930   
28708         True  ...    2   -7.683    0       0.1940        0.2550   
28709        False  ...   11   -8.307    0       0.0946        0.7010   
28710         True  ...   11   -4.107    1       0.0434        0.3110   
28711         True  ...    6   -3.180    0       0.0319        0.2560   

       instrumentalness  liveness  valence    tempo  time_signature  
0              0.000000    0.2590    0.105   88.880               3  
1              0.693000    0.3040    0.210  118.024               4  
2              0.000135    0.0400   

In [ ]:
spotify_clean_df['snapshot_day'] = spotify_clean_df.snapshot_date.dt.day

spotify_clean_df

spotify_id  \
0      0AYt6NMyyLd0rLuvr0UkMH   
1      2V2K1hzCtgj9xAnga9WUTy   
2      1CrwpdodnkXcjPSrg8kw9g   
3      6HmiaxYKrzq3sU2M0eFDg2   
4      4tsVMjM60RNTe9EV5oQ4sQ   
...                       ...   
28707  3xMHXmedL5Rvfxmiar9Ryv   
28708  56y1jOTK0XSvJzVv9vHQBK   
28709  5mjYQaktjmjcMKcUIcqz4s   
28710  7x9aauaA9cu6tyfpHnqDLo   
28711  3rUGC1vUpkDG9CZFHMur1t   

                                                    name  \
0                              Slime You Out (feat. SZA)   
1                       Horns In The Sun - Thakzin Remix   
2      Gangnam Style (feat. DJ Maphorisa & Kabza De S...   
3                                             BETTER NOW   
4                                            Masithokoze   
...                                                  ...   
28707                                            Chaleya   
28708                                 Paint The Town Red   
28709                                          Strangers   
28710                Seven (feat. Latto) (Explicit Ver.)   
28711                                             greedy   

                                                 artists  daily_rank  \
0                                             Drake, SZA          50   
1          DJ Kent, Thakzin, Mo-T, MÖRDA, Brenden Praise          49   
2      Mas Musiq, Daliwonga, DJ Maphorisa, Kabza De S...          48   
3                                                A-Reece          47   
4                                       DJ Stokie, Eemoh          46   
...                                                  ...         ...   
28707  Anirudh Ravichander, Arijit Singh, Shilpa Rao,...           5   
28708                                           Doja Cat           4   
28709                                        Kenya Grace           3   
28710                                   Jung Kook, Latto           2   
28711                                         Tate McRae           1   

       daily_movement  weekly_movement country snapshot_date  popularity  \
0                   0              -28      ZA    2023-10-25          84   
1                 -11                1      ZA    2023-10-25          55   
2                 -15              -13      ZA    2023-10-25          55   
3                   1                3      ZA    2023-10-25          51   
4                   4                4      ZA    2023-10-25          47   
...               ...              ...     ...           ...         ...   
28707              45                0      AE    2023-10-18          84   
28708              46                0      AE    2023-10-18          87   
28709              47                0      AE    2023-10-18          97   
28710              48                0      AE    2023-10-18          98   
28711              49                0      AE    2023-10-18          98   

       is_explicit  ...  loudness mode speechiness  acousticness  \
0             True  ...    -9.243    0      0.0502        0.5080   
1            False  ...    -8.262    0      0.0670        0.0633   
2            False  ...   -12.642    0      0.0560        0.0632   
3             True  ...    -7.524    1      0.1660        0.1300   
4            False  ...   -14.253    0      0.0617        0.0224   
...            ...  ...       ...  ...         ...           ...   
28707        False  ...    -5.280    1      0.0852        0.3930   
28708         True  ...    -7.683    0      0.1940        0.2550   
28709        False  ...    -8.307    0      0.0946        0.7010   
28710         True  ...    -4.107    1      0.0434        0.3110   
28711         True  ...    -3.180    0      0.0319        0.2560   

       instrumentalness  liveness  valence    tempo  time_signature  \
0              0.000000    0.2590    0.105   88.880               3   
1              0.693000    0.3040    0.210  118.024               4   
2              0.000135    0.0400    0.905  111.010               4   
3              0.00003